spring-kafka/src/main/java/org/springframework/kafka/retrytopic/DefaultDestinationTopicResolver.java

```diff
@@ -144,18 +155,44 @@ public DestinationTopic getDestinationTopicByName(String mainListenerId, String
 
 	@Nullable
 	@Override
-	public DestinationTopic getDltFor(String mainListenerId, String topicName) {
-		DestinationTopic destination = getDltOrNoOpsDestination(mainListenerId, topicName);
+	public DestinationTopic getDltFor(String mainListenerId, String topicName, Exception e) {
+		DestinationTopic destination = getDltOrNoOpsDestination(mainListenerId, topicName, e);
 		return destination.isNoOpsTopic()
 				? null
 				: destination;
 	}
 
-	private DestinationTopic getDltOrNoOpsDestination(String mainListenerId, String topic) {
+	private DestinationTopic getDltOrNoOpsDestination(String mainListenerId, String topic, Exception e) {
 		DestinationTopic destination = getNextDestinationTopicFor(mainListenerId, topic);
-		return destination.isDltTopic() || destination.isNoOpsTopic()
-				? destination
-				: getDltOrNoOpsDestination(mainListenerId, destination.getDestinationName());
+		return isMatchingDltTopic(destination, e) || destination.isNoOpsTopic() ?
+			destination :
+			getDltOrNoOpsDestination(mainListenerId, destination.getDestinationName(), e);
+	}
+
+	private static boolean isMatchingDltTopic(DestinationTopic destination, Exception e) {
+		if (!destination.isDltTopic()) {
+			return false;
+		}
+
+		boolean isDltIntendedForCurrentExc = destination.usedForExceptions().stream()
+			.anyMatch(excType -> isDirectExcOrCause(e, excType));
+		boolean isGenericPurposeDlt = destination.usedForExceptions().isEmpty();
+		return isDltIntendedForCurrentExc || isGenericPurposeDlt;
+	}
+
+	private static boolean isDirectExcOrCause(Exception e, Class<? extends Throwable> excType) {
```

Comment:
Is there a use case in which the traversal of the `cause` chain is necessary? Should we enable/disable this via config?

Reply:
I've implemented it this way, because I think it's a natural behaviour, which might be expected by developers. Let's consider a situation, when someone wants to have a separate DLT per business exception. When this exception is wrapped by `TimestampedException` by Spring Kafka, then custom DLT routing won't kick in if we don't traverse causes. That's because mentioned business exception is one level deeper than the `TimestampedException`, it's its cause, which needs to be looked for.

If you find this important to provide the option to configure this causes traversing behaviour, then I'm open to implement it. I didn't want to introduce it from the beginning, because I'd like to avoid potentially unnecessary complexity.